### ISSUE TITLE
patch_dilation seems problemetic

### ISSUE THREAD

**Issue by jasonbian97** [ClementPinard/Pytorch-Correlation-extension#50]
Hi @ClementPinard , I find something wrong when I use patch_dilation>1. The output shape looks good to me, but the value is weird to me. Here is an example:
```python
import torch
from spatial_correlation_sampler import SpatialCorrelationSampler

torch.manual_seed(0)

device = "cuda"
batch_size =1 
channel =1
H = 10
W = 10
dtype = torch.float32

B = batch_size
kernel_size = 1
patch = 3
stride = 1
pad = 0
patch_dilation = 2

input1 = torch.randint(1,4,(batch_size,channel,H,W),dtype=dtype,device=device,requires_grad=True)
input2 = torch.randint_like(input1,1,4).requires_grad_(True)

correlation_sampler = SpatialCorrelationSampler(
    kernel_size,
    patch,
    stride,
    pad,
    patch_dilation)

out = correlation_sampler(input1,input2)

print(out.shape)
print(input1[0])
print(input2[0])
print(out[0,:,:,0,0])
```
The output is:
```
tensor([[[2., 1., 2., 2., 3., 1., 3., 3., 3., 1.],
         [3., 2., 2., 3., 3., 1., 2., 3., 3., 2.],
         [1., 1., 3., 2., 3., 2., 1., 1., 2., 2.],
         [3., 2., 3., 2., 3., 1., 1., 3., 1., 1.],
         [2., 3., 3., 2., 1., 2., 3., 1., 2., 3.],
         [1., 1., 2., 3., 3., 1., 1., 2., 2., 1.],
         [3., 2., 1., 3., 3., 1., 3., 1., 3., 2.],
         [3., 3., 2., 2., 3., 1., 2., 2., 1., 2.],
         [1., 1., 3., 1., 3., 3., 1., 2., 3., 1.],
         [3., 2., 2., 1., 1., 1., 1., 1., 2., 2.]]], device='cuda:0',
       grad_fn=<SelectBackward>)
tensor([[[2., 2., 1., 1., 3., 1., 1., 1., 1., 3.],
         [2., 1., 3., 3., 2., 2., 1., 1., 2., 3.],
         [2., 2., 2., 2., 1., 2., 2., 1., 1., 2.],
         [1., 3., 3., 3., 3., 3., 1., 1., 1., 3.],
         [1., 3., 1., 2., 3., 1., 2., 3., 3., 1.],
         [3., 1., 3., 1., 3., 2., 1., 2., 1., 2.],
         [3., 2., 2., 1., 1., 2., 2., 1., 2., 1.],
         [3., 2., 2., 1., 1., 3., 2., 1., 1., 1.],
         [1., 2., 3., 1., 1., 3., 3., 3., 3., 1.],
         [3., 2., 1., 3., 3., 3., 2., 2., 2., 2.]]], device='cuda:0',
       grad_fn=<SelectBackward>)
torch.Size([1, 3, 3, 10, 10])
tensor([[0., 0., 0.],
        [0., 4., 4.],
        [0., 4., 2.]], device='cuda:0', grad_fn=<SelectBackward>)
```
I manually compute the "dilated correlation", and I guess the result of `out[0,:,:,0,0]` should be:
```
0 0 0
0 4 2
0 4 4
```
which is:
```
A[0,0]* Matrix(
B[-2,-2] B[-2,0] B[-2,2]
B[0,-2] B[0,0] B[0,2]
B[-2,-2] B[-2,0] B[-2,2]
)
```

**Comment by ClementPinard** [comment]
`patch_dilation`is only useful when kernel > 1, it acts as the dilation parameter in classic convolutions. What you probably want is stride, which, if set to 2 will have a shift of 2 pixel for each different patch used for correlation.

I agree that the name can be a bit misleading, the rationale being the naming convention is that each correlation operation consists in a patch of size `kernel` , with a dilation parameter `dilation_patch`. If you want more correlation to be computed, you can increase the `patch`parameter to have more patches tested around the poition of zero shift. And if you want the shift to be more than one around the zero shift, you need to increase the `stride`parameter.

**Comment by ClementPinard** [comment]
You forgot the `dilation` parameter in your function call. `dilation` refers to the dilation of correlation kernel (as in classic convolutions) and `dilation_patch` refers to the parameter you actually want to change. I changed your code a bit and now it works :

```python
import torch
from spatial_correlation_sampler import SpatialCorrelationSampler

torch.manual_seed(0)

device = "cuda"
batch_size =1
channel =1
H = 10
W = 10
dtype = torch.float32

B = batch_size
kernel_size = 1
patch = 3
stride = 1
pad = 0
patch_dilation = 2

input1 = torch.randint(1,4,(batch_size,channel,H,W),dtype=dtype,device=device,requires_grad=True)
input2 = torch.randint_like(input1,1,4).requires_grad_(True)

correlation_sampler = SpatialCorrelationSampler(
            kernel_size=kernel_size,
            patch_size=patch,
            stride=stride,
            padding=pad,
            dilation_patch=patch_dilation)

out = correlation_sampler(input1,input2)

print(out.shape)
print(input1[0])
print(input2[0])
print(out[0,:,:,0,0])
```

In the mean time I agree that the exact parameter order should appear somewhere in the README to avoid people to have to look for the code to know how to use the function.

**Comment by jasonbian97** [comment]
Great, it works! Thank you sooo much!

**Comment by jasonbian97** [comment]
When I play around the `dilation` parameters, I find another wired thing:
basically, no mater what I set the `dilation`, the results are the same. 
```python
correlation_sampler = SpatialCorrelationSampler(
            kernel_size=3,
            patch_size=3,
            stride=1,
            padding=0,
            dilation = 2,
            dilation_patch=1)

out = correlation_sampler(input1,input2)

correlation_sampler2 = SpatialCorrelationSampler(
            kernel_size=3,
            patch_size=3,
            stride=1,
            padding=0,
            dilation = 1,
            dilation_patch=1)

out2 = correlation_sampler2(input1,input2)

torch.all(out==out2) # True #This is wired
output = out[0,:,:,0,0]
```
the `output` is:
```
tensor([[13., 24., 23.],
        [20., 32., 32.],
        [22., 38., 40.]], device='cuda:0', grad_fn=<SelectBackward>)
```
however, when I manually compute this (`dilation=2`), `output[0,0]` should be 15 instead of 13. 

Thanks again!

**Comment by ClementPinard** [comment]
Well that's embarassing, just checked the code and the dilation parameter is actually not implemented !
Not sure what made me give up this feature, but in the python code, the option is literally thrown away. I'll try to implement it in the near future, but it should not be too complicated.

Thanks for the catch !

**Comment by jasonbian97** [comment]
Np! Thanks for your work!

Oh, I see the difference between our thinking. I compute 15 like this:
Since we are using `kernel_size=3, padding=0`, so the centre of the first image I guess should be at (1,1) instead of (2,2), and the output spatial size should be 8x8 (with implicit 0-padding during the correlation). From your equation, I guess your output size is 6x6. To be more specific, I compute like this:
```
a[-1,-1] * b[-2,-2] + a[-1,1] * b[-2,0] + a[-1,3] * b [-2,2] + \
a[1,-1] * b[0,-2] + a[1,1] * b[0,0] + a[1,3] * b[0,2] + \
a[3,-1] * b[2,-2] + a[3,1] * b[2,0] + a[3,3] * b[2,2]
```
But I don't think this is an issue. I guess after padding your method with 1, we will get the same result.

**Comment by ClementPinard** [comment]
I updated the codebase and the README, dilation should work now.
Can you test it for me ?

thanks !
